scc/light_client/light_client_state/light_client_state_test.go

```diff
@@ -0,0 +1,299 @@
+		Return([]cert.BlockCertificate{}, nil)
+
+	state := NewState(scc.Committee{})
+	_, err := state.Sync(prov)
+	require.ErrorContains(err, "zero block certificates")
+	want := State{}
+	compareStates(t, &want, state)
+}
+
+func TestLightClientState_Sync_UpdatesOnlyHeadWhen_SyncToCurrentPeriod(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	prov := provider.NewMockProvider(ctrl)
+
+	// setup state with non empty committee
+	key := bls.NewPrivateKey()
+	member := makeMember(key)
+	state := NewState(scc.NewCommittee(member))
+	state.period = 1
+
+	// setup block for period 1.
+	blockNumber := idx.Block(scc.BLOCKS_PER_PERIOD*1 + 1)
+	blockCert := cert.NewCertificate(
+		cert.NewBlockStatement(0, blockNumber, common.Hash{0x1}, common.Hash{0x2}))
+
+	// sigh the block certificate with the committee member
+	err := blockCert.Add(scc.MemberId(0), cert.Sign(blockCert.Subject(), key))
+	require.NoError(err)
+
+	prov.EXPECT().
+		GetBlockCertificates(provider.LatestBlock, uint64(1)).
+		Return([]cert.BlockCertificate{blockCert}, nil)
+
+	_, err = state.Sync(prov)
+	require.NoError(err)
+	want := State{
+		period:     scc.Period(1),
+		committee:  scc.NewCommittee(member),
+		headNumber: blockCert.Subject().Number,
+		headHash:   common.Hash{0x1},
+		headRoot:   common.Hash{0x2},
+	}
+	compareStates(t, &want, state)
+}
+
+func TestLightClientState_Sync_CanNotSyncToPastPeriod(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	prov := provider.NewMockProvider(ctrl)
+
+	// setup block for period 1.
+	expectBlockForPeriod(prov, 1)
+
+	// setup synced to period 2
+	state := State{period: 2}
+	_, err := state.Sync(prov)
+	require.ErrorContains(err, "cannot sync to a previous period")
+}
+
+func TestLightClientState_Sync_ReportsFailedVerificationOfLatestBlock(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	prov := provider.NewMockProvider(ctrl)
+
+	// setup state with non empty committee
+	key := bls.NewPrivateKey()
+	member := makeMember(key)
+	state := NewState(scc.NewCommittee(member))
+	state.period = 1
+
+	// setup unsigned block for period 1.
+	expectBlockForPeriod(prov, 1)
+
+	_, err := state.Sync(prov)
+	require.ErrorContains(err, "insufficient voting power")
+}
+
+func TestLightClientState_Sync_FailsWithNilProvider(t *testing.T) {
+	require := require.New(t)
+	state := NewState(scc.Committee{})
+	_, err := state.Sync(nil)
+	require.ErrorContains(err, "cannot update with nil provider")
+}
+
+func TestLightClientState_Sync_FailsWhenHeadProvidedIsSmallerThanCurrent(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	prov := provider.NewMockProvider(ctrl)
+
+	expectBlockForPeriod(prov, 0)
+
+	state := NewState(scc.NewCommittee())
+	lastBlockOfPeriod := idx.Block(3)
+	state.headNumber = lastBlockOfPeriod
+
+	_, err := state.Sync(prov)
+	require.ErrorContains(err, "invalid block number")
```

Comment:
nit: also here, for a user of the light client it might be confusing to get an `invalid block number` error when the user has not even provided a block number.

Reply:
should the light client then silently do nothing when latest returned by the provider is older than the known head? should it be handled the same way as trying to sync to a previous period ?

Reply:
error message improved, and now attempts to sync to latest known block are simply ignored, not an error. 
https://github.com/0xsoniclabs/sonic/pull/100/commits/1fcc5fd8763b4eaee841fcbd1ec1e1abd95f55b1